Plugins/ExaTrkX/include/Acts/Plugins/ExaTrkX/Tensor.hpp

```diff
@@ -0,0 +1,143 @@
+    os << "CPU";
+  } else {
+    os << "CUDA(" << device.index << ")";
+  }
+  return os;
+}
+
+/// Capture the context of the execution
+struct ExecutionContext {
+  Acts::Device device{Acts::Device::Type::eCPU};
+  std::optional<cudaStream_t> stream;
+};
+
+namespace detail {
+
+/// This class implements the memory management for the Acts::Tensor
+class TensorMemoryImpl {
+ public:
+  using Deleter = std::function<void(void *)>;
+
+  TensorMemoryImpl(std::size_t nbytes, const ExecutionContext &execContext);
+
+  TensorMemoryImpl(const TensorMemoryImpl &) = delete;
+  TensorMemoryImpl(TensorMemoryImpl &&) noexcept;
+
+  TensorMemoryImpl &operator=(const TensorMemoryImpl &) = delete;
+  TensorMemoryImpl &operator=(TensorMemoryImpl &&) noexcept;
+
+  ~TensorMemoryImpl();
+
+  TensorMemoryImpl clone(std::size_t nbytes, const ExecutionContext &to) const;
+
+  void *data() { return m_ptr; }
+  const void *data() const { return m_ptr; }
+
+  Acts::Device device() const { return m_device; }
+
+ private:
+  void moveConstruct(TensorMemoryImpl &&) noexcept;
+
+  void *m_ptr{};
+  Deleter m_deleter;
+  Acts::Device m_device{};
+};
+}  // namespace detail
+
+/// This is a very small, limited class that models a 2D tensor of arbitrary
+/// type. It is move-only, and only possible to create via static factory
+/// functions to ensure lifetime management.
+/// This on purpose does not implement operations such as clone/to-host/to-cuda
+template <typename T>
+class Tensor {
+ public:
+  using Shape = std::array<std::size_t, 2>;
+
+  static Tensor Create(Shape shape, const ExecutionContext &execContext) {
+    detail::TensorMemoryImpl memory(shape[0] * shape[1] * sizeof(T),
+                                    execContext);
+    return Tensor(std::move(memory), shape);
+  }
+
+  /// Clone the tensor, copying the data to the new device
+  /// @param to The {device, stream} to clone to
+  /// @note This is a always a deep copy, even if the source and destination are the
+  /// same device
+  Tensor clone(const ExecutionContext &to) const {
+    auto clonedMemory = m_memory.clone(nbytes(), to);
+    return Tensor(std::move(clonedMemory), m_shape);
+  }
+
+  T *data() { return static_cast<T *>(m_memory.data()); }
+  const T *data() const { return static_cast<const T *>(m_memory.data()); }
+  Shape shape() const { return m_shape; }
+  std::size_t size() const { return m_shape[0] * m_shape[1]; }
+  std::size_t nbytes() const { return size() * sizeof(T); }
+  Acts::Device device() const { return m_memory.device(); }
+
+ private:
+  Tensor(detail::TensorMemoryImpl memory, Shape shape)
+      : m_shape(shape), m_memory(std::move(memory)) {}
+
+  Shape m_shape{};
+  detail::TensorMemoryImpl m_memory;
+};
+
+/// Element-wise sigmoid function for float cpu tensors
+/// @param tensor The tensor to apply the sigmoid function to
+/// @param execContext The execution context
+void sigmoid(Tensor<float> &tensor, std::optional<cudaStream_t> stream = {});
```

Comment:
Where does `cudaStream_t` come from?

Reply:
so either from CUDA headers, or its just a pointer to a forward declared struct. Should be not an issue since it is never used directly but only passed to CUDA API.

Reply:
No, unfortunately CUDA API does not respect ACTS naming conventions :D